lib/oidcstrategy.js

```diff
@@ -752,4 +573,468 @@ Strategy.prototype._shouldLoadUserProfile = function shouldLoadUserProfile(issue
+      } else {
+        // use the claims from id_token
+
+        if (load && self._options._isV2) {
+          // AAD v2 has no userinfo endpoint
+          log.warn(`Azure v2.0 does not have a 'userinfo' endpoint, we will use the claims in id_token for the profile.`);
+        }
+
+        // we are not doing auth code so we set the tokens to null
+        const accessToken = null;
+        const refreshToken = null;
+        const params = null;
+
+        return onProfileLoaded(self, {
+          req,
+          sub,
+          iss,
+          profile: makeProfileObject(jwtClaims, jwtClaimsStr),
+          jwtClaims,
+          accessToken,
+          refreshToken,
+          params,
+        });
+     }
+    });
+  });
+};
+
+/**
+ * handle the response where we only get 'code' in the response
+ *
+ * @params {Object} code
+ * @params {Object} req
+ * @params {Function} next
+ */
+Strategy.prototype._authCodeFlowHandler = function authCodeFlowHandler(code, req, next) {
+  /* we will do the following things in order:
+   * (1) use code to get id_token and access_token
+   * (2) validate the id_token and the access_token received
+   * (3) if user asks for userinfo and we are using AAD v1, then we use access_token to get
+   *     userinfo, then make sure the userinfo has the same 'sub' as that in the 'id_token'
+   */
+  const self = this;
+
+  log.info(`we are in the authorization code flow, where we only got 'code'`);
+  log.info('received code: ', code);
+
+  // use `null` as identifier since we are in the callback phase of OAuth 2.0 Dance already
+  // identifier only has impact on the `authorize` endpoint
+  return self.configure(null, (err, config) => {
+    if (err) {
+      return next(err);
+    }
+
+    const oauth2 = new OAuth2(
+      config.clientID, // consumerKey
+      config.clientSecret, // consumer secret
+      '', // baseURL (empty string because we use absolute urls for authorize and token paths)
+      config.authorizationURL, // authorizePath
+      config.tokenURL, // accessTokenPath
+      {} // customHeaders
+    );
+
+    let callbackURL = config.callbackURL;
+    // options.callbackURL is merged into config object while `setOptions` call
+    if (!callbackURL) {
+      return next(new Error('no callbackURL found'));
+    }
+
+    const parsedCallbackURL = url.parse(callbackURL);
+    if (!parsedCallbackURL.protocol) {
+      // The callback URL is relative, resolve a fully qualified URL from the
+      // URL of the originating request.
+      callbackURL = url.resolve(utils.originalURL(req), callbackURL);
+    }
+
+    return oauth2.getOAuthAccessToken(code, {
+      grant_type: 'authorization_code',
+      redirect_uri: callbackURL,
+      }, (getOAuthAccessTokenError, access_token, refresh_token, params) => {
+        if (getOAuthAccessTokenError) {
+          return next(new InternalOAuthError('failed to obtain access token', getOAuthAccessTokenError));
+        }
+
+        var id_token = params.id_token;
```

Comment:
When using 'code id_token' flow, redeeming code at token endpoint returns a second id_token. We must ensure that the `sub` and `iss` claims in this second token are same as the first token. http://openid.net/specs/openid-connect-core-1_0.html#HybridIDToken2